File: lever/src/main/java/com/pin/lever/RemoteViewsAction.java
package com.pin.lever;

import android.app.Notification;
import android.app.NotificationManager;
import android.appwidget.AppWidgetManager;
import android.graphics.Bitmap;
import android.widget.RemoteViews;

import static android.content.Context.NOTIFICATION_SERVICE;
import static com.pin.lever.Utils.getService;

abstract class RemoteViewsAction extends Action<RemoteViewsAction.RemoteViewsTarget> {
    final RemoteViews remoteViews;
    final int viewId;
    Callback callback;

    private RemoteViewsTarget target;

    RemoteViewsAction(Lever lever, Request data, RemoteViews remoteViews, int viewId,
                      int errorResId, int memoryPolicy, int networkPolicy, Object tag, String key,
                      Callback callback) {
        super(lever, null, data, memoryPolicy, networkPolicy, errorResId, null, key, tag, false);
        this.remoteViews = remoteViews;
        this.viewId = viewId;
        this.callback = callback;
    }

    @Override void complete(Bitmap result, Lever.LoadedFrom from) {
        remoteViews.setImageViewBitmap(viewId, result);
        update();
        if (callback != null) {
            callback.onSuccess();
        }
    }

    @Override void cancel() {
        super.cancel();
        if (callback != null) {
            callback = null;
        }
    }

    @Override public void error(Exception e) {
        if (errorResId != 0) {
            setImageResource(errorResId);
        }
        if (callback != null) {
            callback.onError(e);
        }
    }

    @Override RemoteViewsTarget getTarget() {
        if (target == null) {
            target = new RemoteViewsTarget(remoteViews, viewId);
        }
        return target;
    }

    void setImageResource(int resId) {
        remoteViews.setImageViewResource(viewId, resId);
        update();
    }

    abstract void update();

    static class RemoteViewsTarget {
        final RemoteViews remoteViews;
        final int viewId;

        RemoteViewsTarget(RemoteViews remoteViews, int viewId) {
            this.remoteViews = remoteViews;
            this.viewId = viewId;
        }

        @Override public boolean equals(Object o) {
            if (this == o) return true;
            if (o == null || getClass() != o.getClass()) return false;
            RemoteViewsTarget remoteViewsTarget = (RemoteViewsTarget) o;
            return viewId == remoteViewsTarget.viewId && remoteViews.equals(
                    remoteViewsTarget.remoteViews);
        }

        @Override public int hashCode() {
            return 31 * remoteViews.hashCode() + viewId;
        }
    }

    static class AppWidgetAction extends RemoteViewsAction {
        private final int[] appWidgetIds;

        AppWidgetAction(Lever lever, Request data, RemoteViews remoteViews, int viewId,
                        int[] appWidgetIds, int memoryPolicy, int networkPolicy, String key, Object tag,
                        int errorResId, Callback callback) {
            super(lever, data, remoteViews, viewId, errorResId, memoryPolicy, networkPolicy, tag, key,
                    callback);
            this.appWidgetIds = appWidgetIds;
        }

        @Override void update() {
            AppWidgetManager manager = AppWidgetManager.getInstance(lever.context);
            manager.updateAppWidget(appWidgetIds, remoteViews);
        }
    }

    static class NotificationAction extends RemoteViewsAction {
        private final int notificationId;
        private final String notificationTag;
        private final Notification notification;

        NotificationAction(Lever lever, Request data, RemoteViews remoteViews, int viewId,
                           int notificationId, Notification notification, String notificationTag, int memoryPolicy,
                           int networkPolicy, String key, Object tag, int errorResId, Callback callback) {
            super(lever, data, remoteViews, viewId, errorResId, memoryPolicy, networkPolicy, tag, key,
                    callback);
            this.notificationId = notificationId;
            this.notificationTag = notificationTag;
            this.notification = notification;
        }

        @Override void update() {
            NotificationManager manager = getService(lever.context, NOTIFICATION_SERVICE);
            manager.notify(notificationTag, notificationId, notification);
        }
    }
}

File: lever/src/main/java/com/pin/lever/RemoteViewsAction.kt
package com.pin.lever

import android.app.Notification
import android.app.NotificationManager
import android.appwidget.AppWidgetManager
import android.content.Context
import android.graphics.Bitmap
import android.widget.RemoteViews
import com.pin.lever.Lever.LoadedFrom
import com.pin.lever.RemoteViewsAction.RemoteViewsTarget
import com.pin.lever.Utils.getService

internal abstract class RemoteViewsAction(
    lever: Lever?, data: Request?, val remoteViews: RemoteViews, val viewId: Int,
    errorResId: Int, memoryPolicy: Int, networkPolicy: Int, tag: Any?, key: String?,
    var callback: Callback?
) : Action<RemoteViewsTarget?>(
    lever,
    null,
    data,
    memoryPolicy,
    networkPolicy,
    errorResId,
    null,
    key,
    tag,
    false
) {
    private var viewsTarget: RemoteViewsTarget? = null
    public override fun complete(result: Bitmap?, from: LoadedFrom?) {
        remoteViews.setImageViewBitmap(viewId, result)
        update()
        if (callback != null) {
            callback!!.onSuccess()
        }
    }

    public override fun cancel() {
        super.cancel()
        if (callback != null) {
            callback = null
        }
    }

    public override fun error(e: Exception?) {
        if (errorResId != 0) {
            setImageResource(errorResId)
        }
        if (callback != null) {
            callback!!.onError(e!!)
        }
    }

    public override fun getTarget(): RemoteViewsTarget? {
        if (viewsTarget == null) {
            viewsTarget = RemoteViewsTarget(remoteViews, viewId)
        }
        return viewsTarget
    }

    fun setImageResource(resId: Int) {
        remoteViews.setImageViewResource(viewId, resId)
        update()
    }

    abstract fun update()
    internal class RemoteViewsTarget(val remoteViews: RemoteViews, val viewId: Int) {
        override fun equals(o: Any?): Boolean {
            if (this === o) return true
            if (o == null || javaClass != o.javaClass) return false
            val remoteViewsTarget = o as RemoteViewsTarget
            return viewId == remoteViewsTarget.viewId && remoteViews ==
                    remoteViewsTarget.remoteViews
        }

        override fun hashCode(): Int {
            return 31 * remoteViews.hashCode() + viewId
        }

    }

    internal class AppWidgetAction(
        lever: Lever?,
        data: Request?,
        remoteViews: RemoteViews,
        viewId: Int,
        private val appWidgetIds: IntArray,
        memoryPolicy: Int,
        networkPolicy: Int,
        key: String?,
        tag: Any?,
        errorResId: Int,
        callback: Callback?
    ) : RemoteViewsAction(
        lever, data, remoteViews, viewId, errorResId, memoryPolicy, networkPolicy, tag, key,
        callback
    ) {
        override fun update() {
            val manager = AppWidgetManager.getInstance(lever!!.context)
            manager.updateAppWidget(appWidgetIds, remoteViews)
        }

    }

    internal class NotificationAction(
        lever: Lever?,
        data: Request?,
        remoteViews: RemoteViews,
        viewId: Int,
        private val notificationId: Int,
        private val notification: Notification,
        private val notificationTag: String,
        memoryPolicy: Int,
        networkPolicy: Int,
        key: String?,
        tag: Any?,
        errorResId: Int,
        callback: Callback?
    ) : RemoteViewsAction(
        lever, data, remoteViews, viewId, errorResId, memoryPolicy, networkPolicy, tag, key,
        callback
    ) {
        override fun update() {
            val manager = getService<NotificationManager>(
                lever!!.context,
                Context.NOTIFICATION_SERVICE
            )
            manager.notify(notificationTag, notificationId, notification)
        }

    }

}